File: Samples/Digital signatures/SignWithAwsKms/C#/AwsCertificateGenerator.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;
using Amazon.KeyManagementService;
using Amazon.KeyManagementService.Model;

namespace BitMiracle.Docotic.Pdf.Samples
{
    class AwsCertificateGenerator : X509SignatureGenerator
    {
        private const string SubjectDn = "CN=self-generated cert,OU=Docotic.Pdf samples,O=Bit Miracle";
        private static readonly byte[] SerialNumber = new byte[] { 17, 59, 46 };
        private static readonly DateTimeOffset NotBefore = DateTime.UtcNow;
        private static readonly DateTimeOffset NotAfter = NotBefore.AddYears(2);

        private readonly X509SignatureGenerator m_baseGenerator;
        private readonly string m_keyId;
        private readonly SigningAlgorithmSpec m_signingAlgorithm;

        private AwsCertificateGenerator(string keyId, SigningAlgorithmSpec alg, X509SignatureGenerator baseGenerator)
        {
            m_baseGenerator = baseGenerator;
            m_keyId = keyId;
            m_signingAlgorithm = alg;
        }

        public static X509Certificate2 Generate(GetPublicKeyResponse key, SigningAlgorithmSpec signingAlgorithm)
            => Generate(key, signingAlgorithm, SubjectDn, NotBefore, NotAfter, SerialNumber);

        public static X509Certificate2 Generate(
            GetPublicKeyResponse key,
            SigningAlgorithmSpec signingAlgorithm,
            string subjectDN,
            DateTimeOffset notBefore,
            DateTimeOffset notAfter,
            byte[] serialNumber)
        {
            byte[] pkBytes = key.PublicKey.ToArray();
            HashAlgorithmName hashAlg = GetHashAlgorithmName(signingAlgorithm);
            string keySpecString = key.KeySpec.ToString();

            CertificateRequest certificateRequest;
            X509SignatureGenerator baseGenerator;
            if (keySpecString.StartsWith("ECC"))
            {
                ECDsa ecdsa = ECDsa.Create();
                ecdsa.ImportSubjectPublicKeyInfo(pkBytes, out _);
                certificateRequest = new CertificateRequest(subjectDN, ecdsa, hashAlg);
                baseGenerator = CreateForECDsa(ecdsa);
            }
            else if (keySpecString.StartsWith("RSA"))
            {
                RSA rsa = RSA.Create();
                rsa.ImportSubjectPublicKeyInfo(pkBytes, out _);

                string alg = signingAlgorithm.ToString();
                RSASignaturePadding padding = alg.StartsWith("RSASSA_PKCS1_V1_5")
                    ? RSASignaturePadding.Pkcs1
                    : RSASignaturePadding.Pss;

                certificateRequest = new CertificateRequest(subjectDN, rsa, hashAlg, padding);
                baseGenerator = CreateForRSA(rsa, padding);
            }
            else
            {
                throw new ArgumentException("Cannot determine encryption algorithm for " + keySpecString, nameof(key));
            }

            var generator = new AwsCertificateGenerator(key.KeyId, signingAlgorithm, baseGenerator);
            return certificateRequest.Create(new X500DistinguishedName(subjectDN), generator, notBefore, notAfter, serialNumber);
        }

        public override byte[] GetSignatureAlgorithmIdentifier(HashAlgorithmName hashAlgorithm)
        {
            return m_baseGenerator.GetSignatureAlgorithmIdentifier(hashAlgorithm);
        }

        public override byte[] SignData(byte[] data, HashAlgorithmName hashAlgorithm)
        {
            using var kmsClient = new AmazonKeyManagementServiceClient();
            using var stream = new MemoryStream(data);
            var signRequest = new SignRequest()
            {
                SigningAlgorithm = m_signingAlgorithm.ToString(),
                KeyId = m_keyId,
                MessageType = MessageType.RAW,
                Message = stream,
            };
            SignResponse signResponse = kmsClient.SignAsync(signRequest).GetAwaiter().GetResult();
            return signResponse.Signature.ToArray();
        }

        protected override PublicKey BuildPublicKey()
        {
            return m_baseGenerator.PublicKey;
        }

        private static HashAlgorithmName GetHashAlgorithmName(string signingAlgorithm)
        {
            if (signingAlgorithm.EndsWith("256"))
                return HashAlgorithmName.SHA256;

            if (signingAlgorithm.EndsWith("384"))
                return HashAlgorithmName.SHA384;

            if (signingAlgorithm.EndsWith("512"))
                return HashAlgorithmName.SHA512;

            throw new ArgumentException("Cannot determine hash algorithm for " + signingAlgorithm, nameof(signingAlgorithm));
        }
    }
}

File: Samples/Digital signatures/SignWithAwsKms/VB.NET/AwsCertificateGenerator.vb
Imports Amazon.KeyManagementService
Imports Amazon.KeyManagementService.Model
Imports System.IO
Imports System.Security.Cryptography
Imports System.Security.Cryptography.X509Certificates

Class AwsCertificateGenerator
    Inherits X509SignatureGenerator

    Private Const SubjectDn As String = "CN=self-generated cert,OU=Docotic.Pdf samples,O=Bit Miracle"
    Private Shared ReadOnly SerialNumber As Byte() = New Byte() {17, 59, 46}
    Private Shared ReadOnly NotBefore As DateTimeOffset = Date.UtcNow
    Private Shared ReadOnly NotAfter As DateTimeOffset = NotBefore.AddYears(2)

    Private ReadOnly m_baseGenerator As X509SignatureGenerator
    Private ReadOnly m_keyId As String
    Private ReadOnly m_signingAlgorithm As SigningAlgorithmSpec

    Private Sub New(keyId As String, alg As SigningAlgorithmSpec, baseGenerator As X509SignatureGenerator)
        m_baseGenerator = baseGenerator
        m_keyId = keyId
        m_signingAlgorithm = alg
    End Sub

    Public Shared Function Generate(key As GetPublicKeyResponse, signingAlgorithm As SigningAlgorithmSpec) As X509Certificate2
        Return Generate(key, signingAlgorithm, SubjectDn, NotBefore, NotAfter, SerialNumber)
    End Function

    Public Shared Function Generate(
        key As GetPublicKeyResponse,
        signingAlgorithm As SigningAlgorithmSpec,
        subjectDN As String,
        notBefore As DateTimeOffset,
        notAfter As DateTimeOffset,
        serialNumber As Byte()) As X509Certificate2

        Dim pkBytes As Byte() = key.PublicKey.ToArray()
        Dim hashAlg As HashAlgorithmName = GetHashAlgorithmName(signingAlgorithm)
        Dim keySpecString As String = key.KeySpec.ToString()
        Dim certificateRequest As CertificateRequest
        Dim baseGenerator As X509SignatureGenerator

        Dim tmp As Integer
        If keySpecString.StartsWith("ECC") Then
            Dim ecdsa As ECDsa = ECDsa.Create()
            ecdsa.ImportSubjectPublicKeyInfo(pkBytes, tmp)

            certificateRequest = New CertificateRequest(subjectDN, ecdsa, hashAlg)
            baseGenerator = CreateForECDsa(ecdsa)
        ElseIf keySpecString.StartsWith("RSA") Then
            Dim rsa As RSA = RSA.Create()
            rsa.ImportSubjectPublicKeyInfo(pkBytes, tmp)

            Dim alg As String = signingAlgorithm.ToString()
            Dim padding As RSASignaturePadding = If(alg.StartsWith("RSASSA_PKCS1_V1_5"), RSASignaturePadding.Pkcs1, RSASignaturePadding.Pss)
            certificateRequest = New CertificateRequest(subjectDN, rsa, hashAlg, padding)
            baseGenerator = CreateForRSA(rsa, padding)
        Else
            Throw New ArgumentException("Cannot determine encryption algorithm for " & keySpecString, NameOf(key))
        End If

        Dim generator = New AwsCertificateGenerator(key.KeyId, signingAlgorithm, baseGenerator)
        Return certificateRequest.Create(New X500DistinguishedName(subjectDN), generator, notBefore, notAfter, serialNumber)
    End Function

    Public Overrides Function GetSignatureAlgorithmIdentifier(hashAlgorithm As HashAlgorithmName) As Byte()
        Return m_baseGenerator.GetSignatureAlgorithmIdentifier(hashAlgorithm)
    End Function

    Public Overrides Function SignData(data As Byte(), hashAlgorithm As HashAlgorithmName) As Byte()
        Using kmsClient = New AmazonKeyManagementServiceClient()
            Using stream = New MemoryStream(data)
                Dim signRequest = New SignRequest() With {
                .SigningAlgorithm = m_signingAlgorithm.ToString(),
                .KeyId = m_keyId,
                .MessageType = MessageType.RAW,
                .Message = stream
            }
                Dim signResponse As SignResponse = kmsClient.SignAsync(signRequest).GetAwaiter().GetResult()
                Return signResponse.Signature.ToArray()
            End Using
        End Using
    End Function

    Protected Overrides Function BuildPublicKey() As PublicKey
        Return m_baseGenerator.PublicKey
    End Function

    Private Shared Function GetHashAlgorithmName(signingAlgorithm As String) As HashAlgorithmName
        If signingAlgorithm.EndsWith("256") Then Return HashAlgorithmName.SHA256
        If signingAlgorithm.EndsWith("384") Then Return HashAlgorithmName.SHA384
        If signingAlgorithm.EndsWith("512") Then Return HashAlgorithmName.SHA512
        Throw New ArgumentException("Cannot determine hash algorithm for " & signingAlgorithm, NameOf(signingAlgorithm))
    End Function
End Class
